File: elixir/diffie-hellman/lib/diffie_hellman.ex
defmodule DiffieHellman do
  def generate_private_key(prime_p) do
    Enum.random(2..prime_p - 1)
  end

  def generate_public_key(prime_p, prime_g, private_key) do
    mod_exp(prime_g, private_key, prime_p)
  end

  def generate_shared_secret(prime_p, public_key_b, private_key_a) do
    mod_exp(public_key_b, private_key_a, prime_p)
  end

  defp mod_exp(a, b, m) when m > 0, do: mod_exp(rem(a, m), rem(b, m), m, 1)
  defp mod_exp(a, b, m, r) do
    cond do
      b == 0         -> r
      rem(b, 2) == 1 -> mod_exp(a          , b - 1    , m, rem(a*r, m))
      true           -> mod_exp(rem(a*a, m), div(b, 2), m, r          )
    end
  end
end

File: fsharp/diffie-hellman/DiffieHellman.fs
module DiffieHellman
open System
open System.Numerics

let rndGen = new Random()

let private modExp (bs: BigInteger) (exponent: BigInteger) (modulus: BigInteger) : BigInteger = 
    let mutable result = 1I
    let mutable bs = bs % modulus
    let mutable exponent = exponent
    while exponent > 0I do
        if exponent % 2I = 1I then
            result <- (result * bs) % modulus
        else 
            result <- result
        exponent <- exponent >>> 1
        bs <- (bs * bs) % modulus
    result

let privateKey (primeP: BigInteger) =
    rndGen.Next(2, primeP |> int) |> BigInteger

let publicKey primeP primeG privateKey = 
    modExp primeG privateKey primeP

let secret primeP publicKey privateKey = 
    modExp publicKey privateKey primeP
